### ISSUE TITLE
Create Target Gradient in Train Mode 

### ISSUE THREAD

**Issue by Siyuan89** [JonasGeiping/invertinggradients#6]
Hi @JonasGeiping ,

Thank you for the great work. 

I noticed that in the code base the target gradients are produced in eval mode  (`model.eval()`) as also the examples. Can the attack also work if the target gradients are generated in training mode (`model.train()`)? 

I think this changes the behavior of batch norm layers and might potentially cause issues. Appreciate your inputs.

**Comment by JonasGeiping** [comment]
Hi @Siyuan89 

thank you for your interest. For these experiments we have only considered models in eval mode.
To run the attack against models with batch norms in training mode would require some additional considerations:
* Would users in the conceived FL application send local batch statistics to the server or discard them?
* If local batch statistics are sent, then they are likely helpful for recovery and should be matched by statistics of the reconstructed data
* This pyTorch code needs to be modified so that the reconstructed gradient is computed in training mode, but should not update batch statistics during the reconstruction iterations. This should be checked carefully to make sure that no additional info is leaked to the reconstruction algorithm by mistake.

**Comment by JonasGeiping** [comment]
I'll close this issue then. Let me know if any questions remain.
